crates/circuit/src/dag_circuit.rs

```diff
@@ -3350,7 +3350,7 @@ def _format(operand):
     ///     DAGCircuitError: If replacement operation was incompatible with
     ///     location of target node.
     #[pyo3(signature = (node, op, inplace=false, propagate_condition=true))]
-    fn substitute_node(
+    pub fn substitute_node(
```

Comment:
This is not necessarily something for you to do, but it would be more proper to add a `py_` prefix to the name of this function and expose it to python with the same name, something like this:

```suggestion
    #[pyo3(name = "substitute_node", signature = (node, op, inplace=false, propagate_condition=true))]
    pub fn py_substitute_node(
```

I believe I already did something like that with #13137. We should probably be thinking of adding a Rust-native method to perform node_substitution in a follow-up.